zerver/lib/zulip_update_announcements.py

```diff
@@ -0,0 +1,177 @@
+    with open(zulip_update_announcements_file_path) as zulip_updates:
+        content = zulip_updates.read()
+        match = re.search(r"\*\*Zulip update announcements level (\d+)\*\*", content)
+        assert match is not None
+        return int(match.group(1))
+
+
+def get_zulip_update_announcements_message_for_level(level: int) -> str:
+    with open(zulip_update_announcements_file_path) as zulip_updates:
+        content = zulip_updates.read()
+        current_level = f"**Zulip update announcements level {level}**"
+        previous_level = f"**Zulip update announcements level {level-1}**"
+        # Use regex to find the message between current_level and previous_level.
+        pattern = re.compile(
+            f"{re.escape(current_level)}(.*?)" + f"{re.escape(previous_level)}", re.DOTALL
+        )
+        match = re.search(pattern, content)
+        assert match is not None
+        return match.group(1).strip()
+
+
+def get_realms_behind_zulip_update_announcements_level(level: int) -> QuerySet[Realm]:
+    # Filter out deactivated realms. When a realm is later
+    # reactivated, send the notices it missed while it was deactivated.
+    realms = Realm.objects.filter(
+        Q(zulip_update_announcements_level__isnull=True)
+        | Q(zulip_update_announcements_level__lt=level),
+        deactivated=False,
+    ).exclude(string_id=settings.SYSTEM_BOT_REALM)
+    return realms
+
+
+def internal_prep_group_direct_message_for_old_realm(
+    realm: Realm, sender: UserProfile
+) -> Optional[SendMessageRequest]:
+    administrators = realm.get_human_admin_users()
+    emails = [administrator.email for administrator in administrators]
+    content = get_zulip_update_announcements_message_for_level(level=1)
+    return internal_prep_huddle_message(realm, sender, emails, content)
+
+
+def is_group_direct_message_sent_to_admins_atleast_one_week_ago(realm: Realm) -> bool:
+    level_none_to_one_auditlog = RealmAuditLog.objects.filter(
+        realm=realm,
+        event_type=RealmAuditLog.REALM_PROPERTY_CHANGED,
+        extra_data__contains={
+            RealmAuditLog.OLD_VALUE: None,
+            RealmAuditLog.NEW_VALUE: 1,
+            "property": "zulip_update_announcements_level",
+        },
+    ).first()
+    assert level_none_to_one_auditlog is not None
+    group_direct_message_sent_on = level_none_to_one_auditlog.event_time
+    return timezone_now() - group_direct_message_sent_on > timedelta(days=7)
+
+
+def internal_prep_zulip_update_announcements_stream_messages(
+    current_level: int, latest_level: int, sender: UserProfile, realm: Realm
+) -> List[Optional[SendMessageRequest]]:
+    messages = []
+    topic_name = str(realm.ZULIP_UPDATE_ANNOUNCEMENTS_TOPIC_NAME)
+    stream = realm.zulip_update_announcements_stream
+    assert stream is not None
+    while current_level < latest_level:
+        content = get_zulip_update_announcements_message_for_level(level=current_level + 1)
+        messages.append(
```

Comment:
Call this variable `message_requests`, for readability.

Reply:
Done.